dao-voting/src/main.sw

```diff
@@ -0,0 +1,272 @@
+    #[storage(read, write)]
+    fn deposit() -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+        require(storage.gov_token == msg_asset_id(), Error::NotGovernanceToken);
+        require(msg_amount() > 0, Error::NoAssetsSent);
+
+        let result: Result<Identity, AuthError> = msg_sender();
+        let sender: Identity = result.unwrap();
+
+        let prev_balance = storage.balances.get(sender);
+        let new_balance = prev_balance + msg_amount();
+        storage.balances.insert(sender, new_balance);
+
+        true
+    }
+
+    /// Update the user balance to indicate they have withdrawn governance tokens
+    ///
+    /// # Panics
+    ///
+    /// This functions will panic when:
+    /// - The constructor has not been called to initalize
+    /// - The user tries to withdraw more than their balance
+    #[storage(read, write)]
+    fn withdraw(amount: u64) -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+
+        let result: Result<Identity, AuthError> = msg_sender();
+        let sender: Identity = result.unwrap();
+
+        let prev_balance = storage.balances.get(sender);
+        require(prev_balance >= amount, Error::NotEnoughAssets);
+
+        let new_balance = prev_balance - amount;
+        storage.balances.insert(sender, new_balance);
+
+        // Transfer the asset back to the user
+        match sender {
+            Identity::Address(address) => transfer_to_output(amount, storage.gov_token, address), 
+            Identity::ContractId(address) => force_transfer_to_contract(amount, storage.gov_token, address), 
+        }
+
+        true
+    }
+
+    /// Vote on a given proposal
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor has not been called to initialize
+    /// - The proposal id is out of range
+    /// - The vote amount is 0
+    /// - The vote amount is greater than the users deposited balance
+    /// - The given proposal is expired
+    #[storage(read, write)]
+    fn vote(proposal_id: u64, vote_amount: u64, is_yes_vote: bool) -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+        require(proposal_id < storage.proposal_count, Error::InvalidId);
+        require(vote_amount > 0, Error::VoteAmountCannotBeZero);
+
+        let result: Result<Identity, AuthError> = msg_sender();
+        let sender: Identity = result.unwrap();
+        let sender_balance = storage.balances.get(sender);
+
+        require(sender_balance >= vote_amount, Error::NotEnoughAssets);
+
+        let mut proposal = storage.proposals.get(proposal_id);
+        require(proposal.end_height >= height(), Error::ProposalExpired);
+
+        if (is_yes_vote) {
+            proposal.yes_votes = proposal.yes_votes + vote_amount;
+        } else {
+            proposal.no_votes = proposal.no_votes + vote_amount;
+        };
+
+        storage.balances.insert(sender, sender_balance - vote_amount);
```

Comment:
If there are two proposals active at the same time shouldn't the user have access to all their tokens for each proposal individually? If we remove it from their balance their vote will be decreased in the other proposal. Not sure what the convention in solidity DAO voting is

Reply:
I'm not sure I'll do more research.  My thought behind not allowing a user to use their balance for all proposals was to make them prioritize which proposals they wanted to see included.

Reply:
I looked into how MakerDAO governance works, and it works how you describe above. (You can use the same MKR token to vote in any number of polls/proposals).  I'll create an issue for this and tackle it in a separate PR, unless anyone thinks its necessary to include it in this one.